File: backend/app/schemas/coding.py
"""
Pydantic schemas for the coding challenges and courses module.
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Topics
# ---------------------------------------------------------------------------

class TopicResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    title_kk: Optional[str] = None
    slug: str
    description: Optional[str] = None
    description_kk: Optional[str] = None
    sort_order: int
    icon: Optional[str] = None
    grade_level: Optional[int] = None
    is_active: bool


class TopicWithProgress(TopicResponse):
    """Topic enriched with student progress stats."""
    total_challenges: int = 0
    solved_challenges: int = 0


# ---------------------------------------------------------------------------
# Challenges
# ---------------------------------------------------------------------------

class TestCasePublic(BaseModel):
    """Public view of a test case (hidden tests show only description)."""
    input: str
    expected_output: str
    description: Optional[str] = None
    is_hidden: bool = False


class ChallengeListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    title_kk: Optional[str] = None
    difficulty: str
    points: int
    sort_order: int
    # Student-specific
    status: str = "not_started"  # not_started, attempted, solved


class ChallengeDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    topic_id: int
    title: str
    title_kk: Optional[str] = None
    description: str
    description_kk: Optional[str] = None
    difficulty: str
    points: int
    starter_code: Optional[str] = None
    hints: list = Field(default_factory=list)
    hints_kk: list = Field(default_factory=list)
    test_cases: list[TestCasePublic] = Field(default_factory=list)
    time_limit_ms: Optional[int] = 5000
    # Student-specific
    status: str = "not_started"
    best_submission: Optional["SubmissionResponse"] = None


# ---------------------------------------------------------------------------
# Submissions
# ---------------------------------------------------------------------------

class SubmissionCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=50000)
    tests_passed: int = Field(..., ge=0)
    tests_total: int = Field(..., ge=1)
    execution_time_ms: Optional[int] = None
    error_message: Optional[str] = None


class SubmissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    challenge_id: int
    code: str
    status: str
    tests_passed: int
    tests_total: int
    execution_time_ms: Optional[int] = None
    error_message: Optional[str] = None
    attempt_number: int
    xp_earned: int
    created_at: datetime


# ---------------------------------------------------------------------------
# Stats
# ---------------------------------------------------------------------------

class CodingStats(BaseModel):
    total_solved: int = 0
    total_attempts: int = 0
    total_xp: int = 0
    topics_progress: list[TopicWithProgress] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Courses (Learning Paths)
# ---------------------------------------------------------------------------

class CourseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    title_kk: Optional[str] = None
    description: Optional[str] = None
    description_kk: Optional[str] = None
    slug: str
    grade_level: Optional[int] = None
    total_lessons: int = 0
    estimated_hours: Optional[float] = None
    sort_order: int
    icon: Optional[str] = None
    is_active: bool


class CourseWithProgress(CourseResponse):
    """Course enriched with student progress."""
    completed_lessons: int = 0
    last_lesson_id: Optional[int] = None
    started: bool = False
    completed: bool = False


# ---------------------------------------------------------------------------
# Lessons
# ---------------------------------------------------------------------------

class LessonListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    title_kk: Optional[str] = None
    sort_order: int
    has_challenge: bool = False
    challenge_id: Optional[int] = None
    # Student-specific
    is_completed: bool = False


class LessonDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    course_id: int
    title: str
    title_kk: Optional[str] = None
    sort_order: int
    theory_content: str
    theory_content_kk: Optional[str] = None
    starter_code: Optional[str] = None
    challenge_id: Optional[int] = None
    challenge: Optional[ChallengeDetail] = None
    # Student-specific
    is_completed: bool = False


class LessonCompleteRequest(BaseModel):
    """Sent when student finishes reading theory (challenge completion is separate)."""
    pass


class LessonCompleteResponse(BaseModel):
    lesson_id: int
    course_id: int
    completed_lessons: int
    total_lessons: int
    course_completed: bool = False

File: student-app/src/lib/api/coding.ts
/**
 * Student Coding Challenges API
 */

import { apiClient } from './client';

// =============================================================================
// Types
// =============================================================================

export interface CodingTopic {
  id: number;
  title: string;
  title_kk: string | null;
  slug: string;
  description: string | null;
  description_kk: string | null;
  sort_order: number;
  icon: string | null;
  grade_level: number | null;
  is_active: boolean;
  total_challenges: number;
  solved_challenges: number;
}

export interface ChallengeListItem {
  id: number;
  title: string;
  title_kk: string | null;
  difficulty: 'easy' | 'medium' | 'hard';
  points: number;
  sort_order: number;
  status: 'not_started' | 'attempted' | 'solved';
}

export interface TestCase {
  input: string;
  expected_output: string;
  description: string | null;
  is_hidden: boolean;
}

export interface CodingSubmission {
  id: number;
  challenge_id: number;
  code: string;
  status: 'passed' | 'failed' | 'error' | 'timeout';
  tests_passed: number;
  tests_total: number;
  execution_time_ms: number | null;
  error_message: string | null;
  attempt_number: number;
  xp_earned: number;
  created_at: string;
}

export interface ChallengeDetail {
  id: number;
  topic_id: number;
  title: string;
  title_kk: string | null;
  description: string;
  description_kk: string | null;
  difficulty: 'easy' | 'medium' | 'hard';
  points: number;
  starter_code: string | null;
  hints: string[];
  hints_kk: string[];
  test_cases: TestCase[];
  time_limit_ms: number | null;
  status: 'not_started' | 'attempted' | 'solved';
  best_submission: CodingSubmission | null;
}

export interface SubmitSolutionRequest {
  code: string;
  tests_passed: number;
  tests_total: number;
  execution_time_ms?: number;
  error_message?: string;
}

export interface CodingStats {
  total_solved: number;
  total_attempts: number;
  total_xp: number;
  topics_progress: CodingTopic[];
}

// =============================================================================
// API Functions
// =============================================================================

const BASE = '/students/coding';

export async function listTopics(): Promise<CodingTopic[]> {
  const { data } = await apiClient.get(`${BASE}/topics`);
  return data;
}

export async function listChallenges(
  topicSlug: string
): Promise<ChallengeListItem[]> {
  const { data } = await apiClient.get(
    `${BASE}/topics/${topicSlug}/challenges`
  );
  return data;
}

export async function getChallengeDetail(
  challengeId: number
): Promise<ChallengeDetail> {
  const { data } = await apiClient.get(`${BASE}/challenges/${challengeId}`);
  return data;
}

export async function submitSolution(
  challengeId: number,
  body: SubmitSolutionRequest
): Promise<CodingSubmission> {
  const { data } = await apiClient.post(
    `${BASE}/challenges/${challengeId}/submit`,
    body
  );
  return data;
}

export async function listSubmissions(
  challengeId: number
): Promise<CodingSubmission[]> {
  const { data } = await apiClient.get(
    `${BASE}/challenges/${challengeId}/submissions`
  );
  return data;
}

export async function getCodingStats(): Promise<CodingStats> {
  const { data } = await apiClient.get(`${BASE}/stats`);
  return data;
}

// =============================================================================
// Course Types
// =============================================================================

export interface CodingCourse {
  id: number;
  title: string;
  title_kk: string | null;
  description: string | null;
  description_kk: string | null;
  slug: string;
  grade_level: number | null;
  total_lessons: number;
  estimated_hours: number | null;
  sort_order: number;
  icon: string | null;
  is_active: boolean;
  completed_lessons: number;
  last_lesson_id: number | null;
  started: boolean;
  completed: boolean;
}

export interface LessonListItem {
  id: number;
  title: string;
  title_kk: string | null;
  sort_order: number;
  has_challenge: boolean;
  challenge_id: number | null;
  is_completed: boolean;
}

export interface LessonDetail {
  id: number;
  course_id: number;
  title: string;
  title_kk: string | null;
  sort_order: number;
  theory_content: string;
  theory_content_kk: string | null;
  starter_code: string | null;
  challenge_id: number | null;
  challenge: ChallengeDetail | null;
  is_completed: boolean;
}

export interface LessonCompleteResponse {
  lesson_id: number;
  course_id: number;
  completed_lessons: number;
  total_lessons: number;
  course_completed: boolean;
}

// =============================================================================
// Course API Functions
// =============================================================================

export async function listCourses(): Promise<CodingCourse[]> {
  const { data } = await apiClient.get(`${BASE}/courses`);
  return data;
}

export async function listLessons(
  courseSlug: string
): Promise<LessonListItem[]> {
  const { data } = await apiClient.get(
    `${BASE}/courses/${courseSlug}/lessons`
  );
  return data;
}

export async function getLessonDetail(
  lessonId: number
): Promise<LessonDetail> {
  const { data } = await apiClient.get(`${BASE}/courses/lessons/${lessonId}`);
  return data;
}

export async function completeLesson(
  lessonId: number
): Promise<LessonCompleteResponse> {
  const { data } = await apiClient.post(
    `${BASE}/courses/lessons/${lessonId}/complete`
  );
  return data;
}
